File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/dfs/Problem1325.java
package com.keunwon.algorithm.dfs;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Problem1325 {
    static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int N;
    static int M;
    static List<Integer>[] list;

    static int[] ans;

    public static void main(String[] args) throws IOException {
        input();

        for (int i = 1; i <= N ; i++) {
            dfs(i, new boolean[N + 1]);
        }

        int max = Integer.MIN_VALUE;
        for (int i = 1; i <= N; i++) {
            max = Math.max(max, ans[i]);
        }

        for (int i = 1; i <= N; i++) {
            if (max == ans[i]) {
                bw.write(i + " ");
            }
        }

        bw.flush();
        br.close();
    }

    static void dfs(int x, boolean[] visited) {
        visited[x] = true;

        for (Integer num : list[x]) {
            if (!visited[num]) {
                ans[num]++;
                dfs(num, visited);
            }
        }
    }

    static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        list = new ArrayList[N + 1];
        ans = new int[N + 1];

        for (int i = 0; i <= N; i++) {
            list[i] = new ArrayList<>();
        }

        for (int i = 1; i <= M; i++) {
            st = new StringTokenizer(br.readLine());
            int num1 = Integer.parseInt(st.nextToken());
            int num2 = Integer.parseInt(st.nextToken());

            list[num1].add(num2);
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem1325.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 효율적인 해킹
 * Level: 실버-1
 **/
class Problem1325 {
    private lateinit var graph: Array<MutableList<Int>>

    fun solution(n: Int, arr: Array<Pair<Int, Int>>): List<Int> {
        this.graph = Array(n + 1) { mutableListOf<Int>() }.apply {
            arr.forEach { (a, b) -> this[b].add(a) }
        }
        var max = 0
        var answer = IntArray(n + 1)

        for (i in 1..n) {
            val count = bfs(i, n)
            max = max.coerceAtLeast(count)
            answer[i] = count
        }
        return (1..n).filter { answer[it] == max }
    }

    private fun bfs(startIndex: Int, n: Int): Int {
        val queue = LinkedList<Int>()
        val visited = BooleanArray(n + 1)
        var count = 0

        queue.offer(startIndex)
        visited[startIndex] = true

        while (queue.isNotEmpty()) {
            val cur = queue.poll()
            count++

            for (next in graph[cur]) {
                if (visited[next]) continue

                visited[next] = true
                queue.offer(next)
            }
        }
        return count
    }
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val (n, m) = br.readLine().split(" ").map { it.toInt() }
        val arr = Array(m) {
            val st = StringTokenizer(br.readLine())
            Pair(st.nextToken().toInt(), st.nextToken().toInt())
        }

        Problem1325().solution(n, arr).also { bw.write("${it.joinToString(" ")}") }
        bw.flush()
    }
}
